ntoskrnl/mm/ARM3/section.c

```diff
@@ -1073,13 +1079,23 @@ MiMapViewInSystemSpace(IN PVOID Section,
     ASSERT(NT_SUCCESS(Status));
 
     /* Get the section size at creation time */
-    SectionSize = ((PSECTION)Section)->SizeOfSection.LowPart;
+    SectionSize = ((PSECTION)Section)->SizeOfSection.QuadPart;
 
-    /* If the caller didn't specify a view size, assume the whole section */
-    if (!(*ViewSize)) *ViewSize = SectionSize;
+    /* If the caller didn't specify a view size, assume until the end of the section */
+    if (!(*ViewSize))
+    {
+        /* Check for overflow first */
+        if ((SectionSize - SectionOffset->QuadPart) > SIZE_T_MAX)
+        {
+            DPRINT1("Section end is too far away from the specified offset.\n");
+            MiDereferenceControlArea(ControlArea);
+            return STATUS_INVALID_VIEW_SIZE;
+        }
+        *ViewSize = SectionSize - SectionOffset->QuadPart;
+    }
 
     /* Check if the caller wanted a larger section than the view */
-    if (*ViewSize > SectionSize)
+    if (SectionOffset->QuadPart + *ViewSize > SectionSize)
```

Comment:
May want to check for integer overflow here too (`SectionOffset->QuadPart + *ViewSize < SectionOffset->QuadPart`)?

Reply:
Right.